Add tox support (#1226)
openstenoproject/plover: doc/developer_guide.md

# Environment setup

You need Python >= 3.6 installed, and you need [tox](https://pypi.org/project/tox/) >= 3.10.

Using tox takes care of all the details of creating and managing an isolated
virtual environment, installing the necessary dependencies, and isolating
testsuite runs.

The command for using tox is: `tox {-e envlist} {-- arguments}`. Use `tox -a
-v` to get a list of available environments.

The same virtual environment is reused by the following tox environments:

> - `tox -e text -- ARGS`: run the testsuite. This is the default environment

Reviewer: This command has a typo
```suggestion
- `tox -e test -- ARGS`: run the testsuite. This is the default environment
```

  when not provided.
- `tox -e setup -- COMMAND`: run `./setup.py COMMAND`.

> - `tox -e launch -- ARGS`: run `./setup.py launch ARGS`.

Reviewer: Maybe we can label this one
```suggestion
- `tox -e launch -- ARGS`: run `./setup.py launch ARGS`. Run Plover from source.
```

# Environment setup

> You need Python >= 3.6 installed, and you need [tox](https://pypi.org/project/tox/) >= 3.10.

Reviewer: May be good to indicate that it can be pip installed, something like this?

```suggestion
You need Python >= 3.6 installed, and you need [tox](https://pypi.org/project/tox/) >= 3.10 (`python3 -m pip install tox`).
```

Author: Yeah, but that's a given, no? It's a link to PyPI. And you may want to use your distribution package.

Reviewer: Only a concern to me as there's always a chance this is someone's first python project, and they aren't familiar with what PyPI is (and assume that it's a program you have to download.)

I think that if we do make a quick bulleted list like I did in other comments, then it wouldn't be necessary to include a command here, or we could just say "available from pip"

Author: If you want. Personally, I'd prefer things not to be as easy: I don't want people to mindlessly copy/paste commands, it's a recipe for disaster.

Reviewer: I understand the principle. I guess the trade-off is just inevitably someone's gonna open a discussion asking "How do I run from source" and another one that's "how to activate tox environment", but maybe that's a feature, not a bug.

Using tox takes care of all the details of creating and managing an isolated
virtual environment, installing the necessary dependencies, and isolating
testsuite runs.

The command for using tox is: `tox {-e envlist} {-- arguments}`. Use `tox -a
-v` to get a list of available environments.

> 

Reviewer: Maybe right here it would be helpful to just state something like

> To set up and activate the tox environment for Plover and install all dependencies, in your cloned repository simply run:
>
> `tox`

It's so easy that I almost didn't trust it.

Reviewer: Alternatively I would put this right at the top, after saying that python and tox are required.

That way, I could see someone who's just trying to run from source would figure it out…

- Install python
- pip install tox
- tox
- Mac and Linux: ./launch.sh
- Windows: ./launch.bat

Or maybe we just need a really brief "Running from Source" for someone who's just looking to get set up as quickly as possible.

Author: Except that won't work, you need to activate the environment before using lanch.sh/launch.bat. Honestly, I almost remove those. And whether you're using the bat or the sh on Windows will depend on your environment: since we requiring bash to be present for creating the distribution, someone might use the git shell.

Reviewer: Fair, so would the real bootstraps be

- Install python
- pip3 install tox
- tox -e launch

?

Reviewer: I'm also cool with removing the launch scripts 👍

Author: OK, I'll remove the launch/test scripts.

The same virtual environment is reused by the following tox environments:

> - `tox -e text -- ARGS`: run the testsuite. This is the default environment
>   when not precised.

Reviewer: ```suggestion
- `tox -e test -- ARGS`: run the testsuite. This is the default environment
  when not provided.
```

- `tox -e setup -- COMMAND`: run `./setup.py COMMAND`.
- `tox -e launch -- ARGS`: run `./setup.py launch ARGS`.
- `tox -e packaging_checks`: run the same packaging checks as the CI (add `--
  -n` to see a dry-run of the exact checks).
- `tox -e plugins_install`: install the distribution plugins (or the specified
  plugins when run with `tox -e plugins_install -- REQS`). Note that this does
  not use the plugins manager for installing.
- `tox -e release_prepare -- NEW_VERSION`: execute all the steps necessary for

>   preparing a new release; patch the version to `NEW_VERSION`  and update

Reviewer: ```suggestion
  preparing a new release; patch the version to `NEW_VERSION` and update
```

  `NEWS.md`, staging all the changes for review.

> - `tox -e release_finalize`: finalize the release; commit the staged changes,

Reviewer: ```suggestion
- `tox -e release_finalize`: finalize the release: commit the staged changes,
```

Author: Above as well?

Reviewer: Yes, full colon to prefix a list 👍

  create an annotated tag, and print the git command necessary for pushing the
  release to GitHub.

> The actual virtual environment lives in `.tox/dev`, and can be "activated" like
> any other virtual environment.

Reviewer: I think that we should provide a link here to explain what `"activated" like any other virtual environment` means, maybe like this?
```suggestion
The actual virtual environment lives in `.tox/dev`, and can be "activated" like
any other [virtual environment.](https://docs.python.org/3/tutorial/venv.html)
```

Author: I'll link to https://virtualenv.pypa.io/en/latest/user_guide.html?highlight=activate#activators.
